gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java

```diff
@@ -2248,6 +2275,67 @@ private boolean canIgnoreExceptionForEmptyObject(
     return false;
   }
 
+  @Override
+  public boolean isHnBucket(URI src) throws IOException {
+    String bucketName = src.getAuthority();
+    Boolean isEnabled = cache.getIfPresent(bucketName);
+    if (isEnabled != null) {
+      return isEnabled;
+    }
+
+    String prefix = src.getPath().substring(1);
+
+    StorageControlClient storageControlClient = lazyGetStorageControlClient();
+    GetStorageLayoutRequest request =
+        GetStorageLayoutRequest.newBuilder()
+            .setPrefix(prefix)
+            .setName(StorageLayoutName.format("_", bucketName))
+            .build();
+
+    try (ITraceOperation to = TraceOperation.addToExistingTrace("getStorageLayout.Hn")) {
+      boolean result =
+          storageControlClient.getStorageLayout(request).getHierarchicalNamespace().getEnabled();
```

Comment:
recommend checking if storageLayout response `.hasHierarchicalNamespace()` before checking getEnabled.

Reply:
When can this return false? The implementation of getHierarchicalNamespace() will never return null.

Reply:
I might be overly cautious here; I pinged Yao in the case it's always set then it's probably fine.

Reply:
Done